src/dnaio/_core.pyx

```diff
@@ -258,122 +266,127 @@ def fastq_iter(file, sequence_class, Py_ssize_t buffer_size):
                 bufstart += 1
                 bufend += 1
                 extra_newline = True
-            else:
-                break
+            elif last_read_position > record_start:  # Incomplete fastq records are present.
+                    if extra_newline:
+                        last_read_position -= 1
+                    lines = buf[record_start:last_read_position].count(b'\n')
+                    raise FastqFormatError(
+                        'Premature end of file encountered. The incomplete final record was: '
+                        '{!r}'.format(
+                            shorten(buf[record_start:last_read_position].decode('latin-1'),
+                                    500)),
+                        line=n_records * 4 + lines)
+            else:  # EOF Reached. Stop iterating.
+                return
 
         # Parse all complete FASTQ records in this chunk
-        pos = 0
         record_start = 0
         while True:
-            # Parse the name (line 0)
-            if c_buf[pos] != b'@':
-                raise FastqFormatError("Line expected to "
-                    "start with '@', but found {!r}".format(chr(c_buf[pos])),
-                    line=n_records * 4)
-            pos += 1
-            while pos < bufend and c_buf[pos] != b'\n':
-                pos += 1
-            if pos == bufend:
+            ### Check for a complete record (i.e 4 newlines are present)
+            # Use libc memchr, this optimizes looking for characters by
+            # using 64-bit integers. See:
+            # https://sourceware.org/git/?p=glibc.git;a=blob_plain;f=string/memchr.c;hb=HEAD
+            # void *memchr(const void *str, int c, size_t n)
+            name_end_ptr = <char *>memchr(c_buf + record_start, 10, <size_t>(bufend - record_start))
+            if name_end_ptr == NULL or name_end_ptr == bufend_ptr:
                 break
-            endskip = 1 if c_buf[pos-1] == b'\r' else 0
-            name_length = pos - endskip - record_start - 1
-            name_encoded = c_buf + record_start + 1
-            # .decode('latin-1') is 50% faster than .decode('ascii')
-            # This is because PyUnicode_DecodeLatin1 is an alias for
-            # _PyUnicode_FromUCS1. Which directly copies the bytes into a
-            # string object. No operations are taking place. With
-            # PyUnicode_DecodeASCII, all characters are checked whether they
-            # exceed 128.
-            name = c_buf[record_start+1:pos-endskip].decode('latin-1')
-
-            pos += 1
-
-            # Parse the sequence (line 1)
-            sequence_start = pos
-            while pos < bufend and c_buf[pos] != b'\n':
-                pos += 1
-            if pos == bufend:
+            name_end = name_end_ptr - c_buf
+            sequence_start = name_end + 1
+            sequence_end_ptr = <char *>memchr(c_buf + sequence_start, 10, <size_t>(bufend - sequence_start))
```

Comment:
I’m not sure, but could bufend - sequence_start ever become negative? I’m just worried about `sequence_start = name_end + 1`. If name_end_ptr points to the last character of the buffer (that is, the previous memchr returned bufend_ptr - 1), then sequence_start will point to the first character *after* the buffer. Or is this not possible?

Reply:
No. Because before that the following check happens: `if name_end_ptr == NULL or name_end_ptr == bufend_ptr:` 
If `name_end_ptr == bufend_ptr` it also means that `name_end == bufend`. In which case we break.
Name_end must therefore be always smaller than bufend. So sequence_start can never point to a character after the buffer.

Having said that, it is good that you mention this, because after checking the qualities newline the check goes thus:
`if qualities_end_ptr == NULL:`. Because it is OK for the FASTQ record to end at bufend. But that means there needs to be a check somewhere to ensure that `next_record_start`, which is `qualities_end+1` does not exceed bufend. That check was removed in my reorganizing of the code. I readded it now.

Reply:
I replaced the `while True` for the inner loop with `while record_start < bufend` to eliminate the possibility of this occurring. I found it more elegant than putting a check at the end of the loop. If there is a `while` statement anyway it is always nice to pair it with a meaningful condition (if the situation calls for it).

Reply:
The issue I see is actually a different one, see comment above.

Reply:
Ah. It seems I did not properly think the `bufend_ptr` trough. Sorry.